Add HTTP headers to cached outputs in RestApiCache

Konamiman opened the pull request:
_This is the fifth of a series of pull requests to implement output caching for the WooCommerce REST API endpoints. Previous: https://github.com/woocommerce/woocommerce/pull/61878, next: https://github.com/woocommerce/woocommerce/pull/61476_

### Changes proposed in this Pull Request:

This pull request extends the RestApiCache trait introduced in https://github.com/woocommerce/woocommerce/pull/61798 to add support for including HTTP headers as part of the cached response. It works as follows:

- If the trait's `get_response_headers_to_include_in_caching` is overriden to return an array of header names, or if an equivalent `include_headers` parameter is passed to `with_cache`, then only the specified headers will be cached. Note that an empty array means that no headers will be cached whatsoever.
- Otherwise (the above returns `false`, which is the default), then all the response headers will be cached except those returned by the `get_response_headers_to_exclude_from_caching` method (defaulting to an empty array), overridable per-endpoint by passing an `exclude_headers` parameter to `with_cache`.
- Once the set of headers to to cache is determined, the following ones are always excluded: `X-WC-Cache`, `Set-Cookie`, `Date`,	`Expires`, `Last-Modified`, `Age`, `ETag`, `Cache-Control`, `Pragma`.
- Finally, the set of headers to cache is filtered via a `woocommerce_rest_api_cached_headers` filter. An attempt to reintroduce the always excluded headers using this filter will trigger a `doing_it_wrong`.

Without overriding any of these trait methods or using any of the equivalent `with_cache` parameters, the default behavior is to cache all headers except the specified fixed set.

### How to test the changes in this Pull Request:

> **⚠️ Testing Prerequisite:** A persistent object cache must be installed in your WordPress instance. If you are testing locally you can use [this simple file-based drop-in plugin](https://gist.github.com/Konamiman/96aeacc7933b5b89cb735c46f02adf89).

1. Add the following snippet to your site, this will create a simple "things" controller for testing the caching trait:

```php
add_action(
	'rest_api_init',
	function () {
		$controller = new class() extends WP_REST_Controller {
			use Automattic\WooCommerce\Internal\Traits\RestApiCache;

			public function __construct() {
				$this->initialize_rest_api_cache();
			}

			private function register_cached_route( $route, $cache_params = array() ) {
				register_rest_route(
					'wc',
					'/' . $route,
					array(
						'methods'             => 'GET',
						'callback'            => empty( $cache_params ) ? $this->with_cache( array( $this, 'get_items' ) ) : $this->with_cache( array( $this, 'get_items' ), $cache_params ),
						'permission_callback' => '__return_true',
					)
				);
			}

			public function register_routes() {
				// Default controller configuration (includes X-WC-Data-1, X-WC-Data-2).
				$this->register_cached_route( 'thing-1' );

				// Override to exclusion mode with custom exclude list (excludes X-WC-Data-2).
				$this->register_cached_route(
					'thing-2',
					array(
						'include_headers' => false,
						'exclude_headers' => array( 'X-WC-Data-2' ),
					)
				);

				// Override to exclusion mode using controller's exclude list (excludes X-WC-Data-4, X-WC-Data-5).
				$this->register_cached_route( 'thing-3', array( 'include_headers' => false ) );

				// Different inclusion list (includes X-WC-Data-1, X-WC-Data-3, X-WC-Data-5).
				$this->register_cached_route( 'thing-4', array( 'include_headers' => array( 'X-WC-Data-1', 'X-WC-Data-3', 'X-WC-Data-5' ) ) );

				// Empty include array (caches NO headers).
				$this->register_cached_route( 'thing-5', array( 'include_headers' => array() ) );

				// exclude_headers is ignored when include_headers is present (caches X-WC-Data-1, X-WC-Data-2, X-WC-Data-3).
				$this->register_cached_route(
					'thing-6',
					array(
						'include_headers' => array( 'X-WC-Data-1', 'X-WC-Data-2', 'X-WC-Data-3' ),
						'exclude_headers' => array( 'X-WC-Data-2' ),
					)
				);
			}

			public function get_items( $request ) {
				$response = new WP_REST_Response(
					array(
						'id'   => 1,
						'name' => 'Test Thing',
					),
					200
				);

				$response->header( 'X-WC-Data-1', 'from-get-items' );
				$response->header( 'X-WC-Data-2', 'from-get-items' );
				$response->header( 'X-WC-Data-3', 'from-get-items' );
				$response->header( 'X-WC-Data-4', 'from-get-items' );
				$response->header( 'X-WC-Data-5', 'from-get-items' );

				return $response;
			}

			protected function get_default_response_entity_type(): ?string {
				return 'thing';
			}

			protected function get_response_headers_to_include_in_caching( WP_REST_Request $request, ?string $endpoint_id = null ): ?array {
				return array( 'X-WC-Data-1', 'X-WC-Data-2' );
			}

			protected function get_response_headers_to_exclude_from_caching( WP_REST_Request $request, ?string $endpoint_id = null ): array {
				return array( 'X-WC-Data-4', 'X-WC-Data-5' );
			}
		};

		$controller->register_routes();
	}
);
```

2. Test the six endpoints by running `curl -i "http://localhost/wp-json/wc/thing-<1 to 6>" `. On the first call for each you should get a cache miss and all the `X-WC-Data-<1 to 5>` headers should be present in the response, on the second call you should get a cache hit and a subset of these headers that depends on the endpoint:

* For `things-1`: headers 1 and 2.
* For `things-2`: headers 1, 3, 4 and 5.
* For `things-3`: headers 1, 2, and 3.
* For `things-4`: headers 1, 3, and 5.
* For `things-5`: none of these headers.
* For `things-6`: headers 1, 2, and 3.

3. Add the following additional snippet:

```php
add_filter(
	'woocommerce_rest_api_cached_headers',
	function ( $headers, $request, $endpoint_id ) {
		$headers = array_diff( $headers, array( 'X-WC-Data-1' ) );
		$headers[] = 'X-WC-Data-3';
		return $headers;
	}, 10, 3
);
```

then run ``curl -i "http://localhost/wp-json/wc/thing-1?foo=bar" ` (query string argument to force a new cached entry) twice, the second time you should get headers 2 and 3 in the response.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Konamiman committed:
Add HTTP header caching capabilities to the RestApiCache trait

Konamiman committed:
Use 'false' instead of 'null' for include_headers to switch to exclude mode

Konamiman committed:
Update woocommerce_rest_api_cached_headers to also allow adding headers.

Konamiman committed:
Add changelog file

Konamiman committed:
Fix  header

albarin commented on plugins/woocommerce/src/Internal/Traits/RestApiCache.php:
```diff
@@ -99,6 +108,30 @@ trait RestApiCache {
+	 *
+	 * @var int
+	 */
+	private static int $cache_ttl = HOUR_IN_SECONDS;
```
Sorry for adding the same comment again 😆  I think this is not used, it was probably reintroduced by mistake.

Konamiman replied:
🤦🏻 Rebasing is tricky. Removed, thanks.

Konamiman committed:
Enforce the always-excluded headers regardless of filter.

If code hooked to woocommerce_rest_api_cached_headers tries to
reintroduce the always-excluded headers (like Set-Cookie and Cache-Control)
a doing it wrong will be triggered and the offending headers will be removed.

Konamiman commented on plugins/woocommerce/src/Internal/Traits/RestApiCache.php:
```diff
@@ -99,6 +108,30 @@ trait RestApiCache {
+		'ETag',
+		'Cache-Control',
+		'Pragma',
+	);
```
Implemented with one extra improvement: an attempt to reintroduce these headers using the filter will throw a `doing_it_wrong`.

albarin reviewed (approved):
Looking good, thanks for the quick fixes!

✅ Confirmed the correct headers are returned in each of the cases.
